woocommerce/woocommerce
# Fix inconsistent behavior for Net Sales in Analytics for single line refund

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR fixes an error that appeared after creating a partial refund. The Net sales total displayed in the Products report did not match the total shown in the Orders report.

Before

![Screen Recording on 2024-06-28 at 04-31-53](https://github.com/woocommerce/woocommerce/assets/1314156/988fd5a5-8308-47ac-b83b-a97acf6d97bc)

After

![Screen Recording on 2024-06-28 at 04-25-18](https://github.com/woocommerce/woocommerce/assets/1314156/0c5c5781-2a7d-4450-bedb-cf75d03de335)


The problem with partial refunds and Analytics is bigger than this PR. The changes here do:
- [This check here](https://github.com/woocommerce/woocommerce/pull/48768/files#diff-d2438118f148f28d98ab96e1a12bdd753e462f4659970939f2fe3e2b87155f59L417) was preventing the order line from being recorded in the `wc_order_product_lookup` table when there were partial refunds. Affecting the Analytics results.
- It also adds an update to run the action to synchronize the orders with missing lines.

The changes here don't cover:

- Total refunds, issue [here](https://github.com/woocommerce/woocommerce/issues/48775).
- Partial order refunds, issue [here](https://github.com/woocommerce/woocommerce/issues/48954).
- As a follow-up, we should clean the `wc_order_product_lookup` table after deleting an order. Similar to what happens [in Orders report](https://github.com/woocommerce/woocommerce/issues/48955).

Closes #48761.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Before checking out this branch.
2. Create some products and orders with multiple lines. Remember that you'll need a product with status: `Completed`.
3. Refund an item. Don't refund every item in the order, just one and in the way shown [here](https://github.com/woocommerce/woocommerce/assets/1314156/0c5c5781-2a7d-4450-bedb-cf75d03de335).
4. Verify that the `Products` report shows a different `Net sales` value than `Revenue` and `Orders`.

<img width="1618" alt="image" src="https://user-images.githubusercontent.com/5581188/166174959-e90b8037-66f7-4386-bfcc-de852dd6cd0e.png">
<img width="1585" alt="image" src="https://user-images.githubusercontent.com/5581188/166174981-47aa2fd8-f8b8-459a-a735-1deca1186379.png">

5. Now checkout this branch and update the WC version in your DB. It will be under `woocommerce_db_version` (running  `wp option update woocommerce_db_version 9.0.0` in wp-cli would do the trick).
6. Go to `/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending` and verify that every action under `Pending` have been triggered.
7. The `Net sales` now should match. The Products report now shows the same `Net sales` value as Revenue and Orders.
9. Repeat steps 2 and 3.
10. Now, the Product report should have included the new order with the partial refunds.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
